Add generic test for classify method of concrete openml providers

shengwangsw opened the pull request:
The abstract test method AbstractProviderModelBaseTest#classifyIndexOfMaxScoresValue() is done, it is able be called by all the test classes that extend it, which were locally fixed.

shengwangsw committed:
openml-utils: Add UT in AbstractProviderModelBaseTest#classifyIndexOfMaxScoresValue()

The method classify was returning opposite value in the feedzai/openml-java project.
In order to improve UTs, the method AbstractProviderModelBaseTest#classifyIndexOfMaxScoresValue() was added in order to test the method #classify for all the providers.

shengwangsw committed:
openml-utils: Update the documentation of the method AbstractProviderModelBaseTest#classifyIndexOfMaxScoresValue()

The documentation before was specifying the class as it was used only for the provider DataRobot.
However, it is an abstract test, hence the documentation should be abstract as well.
Therefore, it documents the method #classify() should return a number which is the index of the list's maximum value obtained from the method #getClassDistribution().

shengwangsw committed:
openml-utils: update imports in class AbstractProviderModelBaseTest

There were imports that are not used.
To improve the code, they are removed.

MiguelSCruz commented on openml-utils/src/test/java/com/feedzai/openml/util/provider/AbstractProviderModelBaseTest.java:
```diff
@@ -78,6 +79,29 @@
+     *
+     * @throws ModelLoadingException If anything goes wrong during loading.
+     * @throws ModelTrainingException If anything goes wrong during training.
+     *
```
This line is not necessary

MiguelSCruz replied:
This line is not necessary

MiguelSCruz commented on openml-utils/src/test/java/com/feedzai/openml/util/provider/AbstractProviderModelBaseTest.java:
```diff
@@ -78,6 +79,29 @@
+
+    /**
+     * Checks the method #classify() if it ensures correctly the index of the list's maximum value
+     * obtained from the method #getClassDistrubution().
```
This explanation is a bit confusing. Can you take a better look at it?

shengwangsw committed:
openml-utils: Update the documentation of the method AbstractProviderModelBaseTest#classifyIndexOfMaxScoresValue()

The documentation is corrected and its explanation is improved.

nmldiegues commented:
Please fix the commit title --- seems to have lost its format somehow

nmldiegues commented:
Also, the commit title does not make much sense. You want to, instead, mention something that's generic and applicable to the openml-api (Which is where you are doing this).
Something like: Add generic test for classify method of concrete openml providers

nmldiegues commented on openml-utils/src/test/java/com/feedzai/openml/util/provider/AbstractProviderModelBaseTest.java:
```diff
@@ -78,6 +79,29 @@
 
+
+    /**
+     * Checks the method #classify() if it ensures that the providers classify
```
"if it" -> "to ensure"

nmldiegues reviewed (commented):
Just a typo; otherwise LGTM

shengwangsw committed:
openml-utils: Add generic test for classify method of concrete openml providers

The documentation of the method AbstractProviderModelBaseTest#classifyIndexOfMaxScoresValue() had a weak description.

This was corrected.

shengwangsw committed:
openml-utils: Update the documentation of the method classifyIndexOfMaxScoresValue

There was an extra 's'.
Remove the extra 's'.

shengwangsw commented:
I think all provided improvements were done.

nmldiegues commented on openml-utils/src/test/java/com/feedzai/openml/util/provider/AbstractProviderModelBaseTest.java:
```diff
@@ -25,6 +25,7 @@
 import com.feedzai.openml.provider.model.MachineLearningModelLoader;
 import com.feedzai.openml.util.algorithm.MLAlgorithmEnum;
 import com.google.common.collect.ImmutableList;
+import com.google.common.primitives.Doubles;
```
This import is unused, right? Please clean it up

shengwangsw replied:
This is used in the method AbstractProviderModelBaseTest#classifyIndexOfMaxScoresValue() to obtain the index of maximum value from a list of double, e.g. _Doubles.asList(scores).indexOf(maxScore)_.

But I just reviewed the code, and there is a method that uses the ArrayUtils. For this reason I will remove this import and update the method mentioned above to use the following _Arrays.asList(ArrayUtils.toObject(scores)).indexOf(maxScore)_.

shengwangsw committed:
openml-utils: Update the test method to classify()

The usage of Doubles from google primitive is unnecessary.

Update to use ArrayUtils, which is used in other method as well.

nmldiegues commented:
@Shenguito  what's preventing this from being merged + back-ported + released?
@ptraca  ?

ptraca commented on openml-utils/src/test/java/com/feedzai/openml/util/provider/AbstractProviderModelBaseTest.java:
```diff
@@ -78,6 +78,29 @@
+     * Checks the method #classify() to ensure that the providers classify
+     * correctly the index of the maximum value of the scores list.
+     *
+     * @throws ModelLoadingException  If anything goes wrong during loading.
```
I don't think we need to document the @Throws clause in test methods.

shengwangsw replied:
I shared the same opinion, I just followed the pattern of the other methods.
I will change it then.

ptraca commented on openml-utils/src/test/java/com/feedzai/openml/util/provider/AbstractProviderModelBaseTest.java:
```diff
@@ -78,6 +78,29 @@
 
+
+    /**
+     * Checks the method #classify() to ensure that the providers classify
```
It should be something like

```java
    /**
     * Verifies that the {@link ClassificationMLModel#classify(Instance)} " returns the index of the greatest value in 
     * the class probability distribution produced by the calling 
     * {@link ClassificationMLModel#getClassDistribution(Instance)} on the model
     * 
     * @see ClassificationMLModel
     */

```

shengwangsw replied:
Right. I will update it.

ptraca commented on openml-utils/src/test/java/com/feedzai/openml/util/provider/AbstractProviderModelBaseTest.java:
```diff
@@ -78,6 +78,29 @@
+     * @throws ModelTrainingException If anything goes wrong during training.
+     */
+    @Test
+    public void classifyIndexOfMaxScoresValue() throws ModelLoadingException, ModelTrainingException {
```
I think it should be called `canGetClassDistributionMaxValueIndex()`
Also we don't need the specific types of Exceptions, the super type will do.


```java
    @Test
    public void classifyIndexOfMaxScoresValue() throws Exception {
```

shengwangsw replied:
Ok, I will change it.

ptraca commented on openml-utils/src/test/java/com/feedzai/openml/util/provider/AbstractProviderModelBaseTest.java:
```diff
@@ -78,6 +78,29 @@
+    @Test
+    public void classifyIndexOfMaxScoresValue() throws ModelLoadingException, ModelTrainingException {
+
+        final M model = getFirstModel();
```
if you space it like this reads a lot better.

```java
 ...
        final M model = getFirstModel();
        final Instance instance = getDummyInstance();
        final double[] scores = model.getClassDistribution(instance);

        final int classificationIndex = model.classify(instance);
        
        final double maxScore = Arrays.stream(scores).max().getAsDouble();

        assertThat(Arrays.asList(ArrayUtils.toObject(scores)).indexOf(maxScore))
                .as("The index of maximum value")
                .isEqualTo(classificationIndex);

```

shengwangsw replied:
Indeed it seems better.

ptraca commented:
After reviewing the commits in `feedzai-openml` and `feedzai-openml-java`, It seems to me that there isn't any good way to test the H2O model provider implementation on `ClassificationMLModel#classify(Instance)` method. Given that the method isn't used by the platform we should mark it as @deprecated.

The unit test in `AbstractProviderModelBaseTest` can remain for now to validate those implementations that can be currently tested but @override the H20 implementation in the `feedzai-openml-java` repository and provide a dummy implementation.

Don't forget to explain in the comment the way of the empty test implementation.

shengwangsw committed:
openml-utils: Review the UT method to classify()

The Javadoc, name, exception and structure of the UT method can be improved.

Improve the UT test by change the method name, javadoc, exceptions and structure.

shengwangsw committed:
openml-utils: Update the abstract UT to be a protected method to be used for all providers UT

Openml python and R provides models that fails on this test.

Create a method which provides the possibility for each provider to use their model in the UT.

shengwangsw committed:
openml-utils: Deprecate the method ClassificationMLModel#classify()

The method classify aims to get the highest value from the class probabilities distribution obtained from the method #getClassDistribution.
Since we already have the class probabilities distribution, we do not need a extra method to obtained the highest value of the distribution.

Deprecate the method #classify().

ptraca commented:
You forgot to deprecate the `classify` method in the `ClassificationMLModel` interface.

@see `https://docs.oracle.com/javase/8/docs/technotes/guides/javadoc/deprecation/deprecation.html`for examples.

shengwangsw commented:
My bad, I forgot to push.

ptraca commented:
LGTM. good to merge.

ptraca commented:
Don't forget to 

- after releasing a version o the lib to back-port it to the active branchs
- Add the back-port versions and commit SHA's to the merge to main commit
- Update the ticket status and affected version fields

This applies to this library and and the dependents (java/python/r, openml internal, etc)
